pkg/virt-controller/watch/pool/pool.go

```diff
@@ -832,6 +838,68 @@ func (c *Controller) scale(pool *poolv1.VirtualMachinePool, vms []*virtv1.Virtua
 	return nil, false
 }
 
+func (c *Controller) getUnavailableVMICount(vms []*virtv1.VirtualMachine) (int, error) {
+	unavailableCount := 0
+	for _, vm := range vms {
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, err := c.vmiStore.GetByKey(vmiKey)
+		if err != nil {
+			return 0, err
+		}
+		if !exists {
+			vmi, err := c.clientset.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+			if err != nil {
+				return 0, err
+			}
+			obj = vmi
+		}
+		vmi := obj.(*virtv1.VirtualMachineInstance)
+		if vmi.DeletionTimestamp != nil || vmi.Status.Phase == virtv1.Failed || vmi.Status.Phase == virtv1.Unknown {
+			unavailableCount++
+			continue
+		}
+		readyCondition := controller.NewVirtualMachineInstanceConditionManager().GetCondition(vmi, virtv1.VirtualMachineInstanceReady)
+		if readyCondition == nil || readyCondition.Status != k8score.ConditionTrue {
+			unavailableCount++
+		}
+	}
+	return unavailableCount, nil
+}
+
+func (c *Controller) handleUnhealthyVMIs(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) error {
+	notReadyVMs := c.filterNotReadyVMs(vms)
+
+	for _, vm := range notReadyVMs {
+		vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
+		obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
+		if exists {
+			vmi := obj.(*virtv1.VirtualMachineInstance)
+			if vmi.DeletionTimestamp != nil {
+				continue
+			}
+			updateType, err := c.isOutdatedVMI(vm, vmi)
+			if err != nil {
+				return err
+			}
+			if updateType == proactiveUpdateTypeRestart {
```

Comment:
You could construct a `metav1.DeleteOptions{}` in the conditionals and then have a single block that does the delete, effectively de-duping the code.